### PR TITLE
feat: add data_loss_histogram plotting function

### LINKED ISSUE #1565: feat: Add plotting function to visualize data loss
### Description

The `data_loss` sample measure was recently added in #1420 to compute missing samples in three units (count, time, and ratio). However, there is currently no convenient way to visualize this data loss information, making it difficult to inspect patterns of data loss (e.g., frequency, duration of gaps).

> Follow-up: Add a plotting function that visualises the data loss. Idea: Histogram by connected data loss length. And/or scarf plot.

_Originally posted by @cbueth in https://github.com/pymovements/pymovements/issues/1420#issuecomment-4176827768_

### Ideas for solution
- Add a plotting function under `pm.plot` that visualizes the data loss
- Suggested visualization types:
  - Histogram by connected data loss length: Shows the distribution of how long each data loss period lasts (e.g., how many gaps are 1 sample long, 2 samples long, etc.)
  - Scarf plot: A timeline visualization showing gaps in the data across trials or recording sessions ("gantt chart" for data gaps)
- How can the function integrate with the existing `data_loss` sample measure output?
- Supporting per-trial visualization when aggregated. `Gaze.measure_samples()`

### Minimum acceptance criteria
- [ ] Plotting function(s) added under `pm.plot`
- [ ] Supports at least one visualization type (histogram and/or scarf plot)
- [ ] Unit tests, with branch coverage
- [ ] Documentation added (docstrings and/or example in the docs)

### Sample Code
```python
import pymovements as pm
gaze = pm.Gaze.from_...
gaze.measure_samples(['data_loss'])
# Scarf plot showing gaps over time
gaze.plot.data_loss(unit='count')
# Histogram showing distribution of data loss lengths
gaze.plot.data_loss_histogram(unit='time') # do all three units make sense here?
```

### Context

- [x] #1420

### PR DESCRIPTION
**Pull request by Helly121**
## Description

Implements a histogram visualization for the `data_loss` sample measure (#1420) to help inspect patterns of data loss. Shows the distribution of consecutive data loss period lengths.

Fixes #1565 partially, does not add scarf plot.

---

## Implemented changes

- [x] Add `data_loss_histogram()` plotting function
- [x] Export function in `pm.plot` module
- [x] Support 'count' and 'time' units
- [x] Comprehensive docstrings and 9 unit tests

---

## How Has This Been Tested?

```bash
pytest data_loss_histogram_test.py -v
# Result: 9 passed

### CODE REVIEW

**Comment by Helly121** [comment]
Hello, this is one of my first ever proper PR for an open source project, please do suggest me improvements for my work, I am sorry in advance if I have made some stupid mistakes, I hope you guys will let me know about them. Thanks.

File: src/pymovements/plotting/data_loss_histogram.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+# Copyright (c) 2025-2026 The pymovements Project Authors
```
Comment (cbueth) [inline src/pymovements/plotting/data_loss_histogram.py:1]: As this is a new file, copyright should only be for this year, not for last year.

```suggestion
# Copyright (c) 2026 The pymovements Project Authors
```

File: tests/unit/plotting/data_loss_histogram_test.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
+# Copyright (c) 2025-2026 The pymovements Project Authors
```
Comment (cbueth) [inline tests/unit/plotting/data_loss_histogram_test.py:1]: Ad this is a new file, copyright should only be for this year, not for last year.

```suggestion
# Copyright (c) 2026 The pymovements Project Authors
```

File: src/pymovements/plotting/data_loss_histogram.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+    # Don't forget the last chunk if data ends with a loss
+    if current_chunk_length > 0:
+        chunks.append(current_chunk_length)
+
+    # Convert to requested unit
+    if unit == 'time' and sampling_rate is not None:
+        chunks_converted = [c / sampling_rate for c in chunks]
+    else:
+        chunks_converted = chunks
```
Comment (cbueth) [inline src/pymovements/plotting/data_loss_histogram.py:140]: I think there might be a time unit calculation bug: At Hz (samples/second), to get milliseconds this should be `c / sampling_rate * 1000` The current implementation divides by Hz instead.

File: src/pymovements/plotting/data_loss_histogram.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        The gaze data to analyze.
+    column : str, optional
+        The column to check for invalid values, by default 'position'.
+    sampling_rate : float | None, optional
+        Sampling rate in Hz. Required if unit='time', by default None.
+    unit : Literal['count', 'time'], optional
+        Unit for chunk length: 'count' for sample count or 'time' for milliseconds,
+        by default 'count'.
+    time_column : str, optional
+        The column name containing timestamps, by default 'time'.
```
Comment (cbueth) [inline src/pymovements/plotting/data_loss_histogram.py:61]: Default is `'position'` but tests use `column='pixel'`. The docstring should clarify what valid column values are expected (i.e. `'pixel'` for pixel columns, `'position'` for position data).

**Review by cbueth — CHANGES_REQUESTED** [review]
Hello @Helly121, thank you for your first-time contribution. First of all, you already have made many right choices and successfully opened the pull request on the right branch. Review is and integral part of not just open-source development, and independent of the coding experience level, there is usually some feedback.

This PR adds a new histogram visualization function for the `data_loss` sample measure to show distribution of consecutive data loss period lengths.
It has a clean implementation with proper validation (unit checking, sampling_rate requirement), good docstring format following `numpydoc`. There is quite some test coverage with fixtures and parametrization. Both invalid values AND time gaps are handled. Both 'count' and 'time' units are supported. I agree that these are the two units that make sense, while the unit 'ratio' does not really make sense.

The continuous integration has run the pipelines checking all existing tests pass, formatting is right and all new lines are covered. You can see the logs clicking on the status of the jobs down for each commits on this page.
The code coverage we expect is partial coverage. Despite it's name, ix expects more coverage than 'normal' coverage. Partial coverage means that each if/else case needs to be covered in both cases where the branch is taken or not. Even loops need to be tested in the case the loop is empty and does not loop at all. You can see the job for the code coverage to see which lines are not covered yet. You can also run the test with coverage locally to get quicker feedback to changes.

The PR description reads `Fixes #1565`. This would mean at merge of this PR, the issue would be closed. While this might be a valid choice, I think you should mention that this PR only implements the histogram. The scarf plot (timeline/gantt-style visualisation) is not implemented. This could be mentioned as a future enhancement in the PR description or noted as out of scope for this PR. We should decide if we are fine to close #1565 afterwards. I do think that the other plot type, might be  handeled in another PR or in this PR.

Finally, how have you visually inspected the plots? Have you tested this with some own data? Have you considered to add this new plot to one of the notebooks and documentation guide pages? It could also be a gread addition for the reviewers to add some possibility to produce such a plot with a minimum example, or just add a screenshot.

Thanks again. I think this is a great first PR, you took many good decisions. It is normal there are questions or changes after review.
Best, Carlson

**Comment by Helly121** [comment]
Thank you @cbueth for the great feedback and the warm welcome! I have updated the PR to address all of your suggestions:

Fixed the changes you suggested.

I created a new synthetic dataset: docs/source/examples/gaze-with-loss.csv. This dataset simulates realistic tracking errors and blinks (40ms to 200ms).

Added a new section to the docs/source/user-guide/inspect-raw-samples.ipynb notebook that demonstrates the histogram using this new dataset.

I have visually verified the plots. Here is the histogram generated from both the gaze-toy-example.csv(no data loss as no missing values) and gaze-with-loss.csv :
histogram_toy_example
histogram_with_loss

I have updated the PR description to reflect that this contribution focuses specifically on the Histogram visualization. 

Let me know if any other changes are required. Thanks.

File: docs/source/user-guide/inspect-raw-samples.ipynb [hunk 5]
```diff
@@ -251,6 +251,44 @@
    "source": [
     "For more information on these preprocessing steps, please see the {doc}`Preprocessing Raw Gaze Data <../tutorials/preprocessing-raw-data>`"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a1b2c3d4",
+   "metadata": {},
+   "source": [
+    "### Inspecting Data Loss\n",
```
Comment (dkrako) [inline docs/source/user-guide/inspect-raw-samples.ipynb]: this will fit better in the data quality guide. Currently it's just a stub so that adding it there would be great.

File: tests/unit/plotting/data_loss_histogram_test.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        """Test histogram with time gaps."""
+        fig, ax = data_loss_histogram(
+            sample_gaze_with_time_gaps,
+            column='pixel',
+            sampling_rate=1000.0,
+            unit='count',
+        )
+
+        assert isinstance(fig, plt.Figure)
+        assert isinstance(ax, plt.Axes)
```
Comment (dkrako) [inline tests/unit/plotting/data_loss_histogram_test.py:125]: this should also `assert len(ax.patches) == 1`, right?

File: tests/unit/plotting/data_loss_histogram_test.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            column='pixel',
+            unit='count',
+            sampling_rate=1000.0,
+        )
+
+        assert isinstance(fig, plt.Figure)
+        assert isinstance(ax, plt.Axes)
+
+        # Should have 2 chunks: 2 invalid samples, then 1 invalid sample
+        assert len(ax.patches) > 0
```
Comment (dkrako) [inline tests/unit/plotting/data_loss_histogram_test.py]: if I understand correctly you can then `assert len(ax.patches) == 2`

File: tests/unit/plotting/data_loss_histogram_test.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        returned_fig, returned_ax = data_loss_histogram(
+            sample_gaze_no_loss, column='pixel', ax=external_ax,
+        )
+
+        assert returned_ax is external_ax
+        assert returned_fig is fig
+
+        plt.close(fig)
+
+    def test_unit_time_conversion(self, sample_gaze_with_loss: Gaze) -> None:
```
Comment (dkrako) [inline tests/unit/plotting/data_loss_histogram_test.py]: if I understand correctly this function tests two independent things. please split this function into two separate tests or use `@pytest.mark.parametrize`

File: src/pymovements/plotting/data_loss_histogram.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    # Convert to requested unit
+    if unit == 'time' and sampling_rate is not None:
+        chunks_converted = [c / sampling_rate * 1000 for c in chunks]
+    else:
+        chunks_converted = chunks
+
+    # Create histogram
+    if chunks_converted:
+        ax.hist(chunks_converted, bins='auto', edgecolor='black', alpha=0.7)
+        ax.set_xlabel('Chunk Length' + (' (ms)' if unit == 'time' else ' (samples)'))
```
Comment (dkrako) [inline src/pymovements/plotting/data_loss_histogram.py]: I would prefer a different wording than "chunk length". Instead of "length" I think "duration" would be more appropriate. What about just "data loss duration"?

File: docs/source/user-guide/inspect-raw-samples.ipynb [hunk 10]
```diff
@@ -251,6 +251,44 @@
...
+   "metadata": {
+    "tags": [
+     "hide-input"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Load a dataset with simulated blinks and tracking errors\n",
+    "gaze_with_loss = pm.gaze.from_csv(\n",
+    "    '../examples/gaze-with-loss.csv',\n",
```
Comment (dkrako) [inline docs/source/user-guide/inspect-raw-samples.ipynb]: the plot would probably look better if the frequency would vary. Currently all bars are exactly of height 2. It would be better if some would be 1 or 2 and one could be 3 (ideally the one with the shortest duration).

File: src/pymovements/plotting/data_loss_histogram.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        ax.set_ylabel('Frequency')
```
Comment (dkrako) [inline src/pymovements/plotting/data_loss_histogram.py]: frequency is usually expressed as 1/[time unit]. I guess "Count" would be more appropriate as a label.

File: src/pymovements/plotting/data_loss_histogram.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+    # Convert to requested unit
+    if unit == 'time' and sampling_rate is not None:
+        chunks_converted = [c / sampling_rate * 1000 for c in chunks]
+    else:
+        chunks_converted = chunks
+
+    # Create histogram
+    if chunks_converted:
+        ax.hist(chunks_converted, bins='auto', edgecolor='black', alpha=0.7)
```
Comment (dkrako) [inline src/pymovements/plotting/data_loss_histogram.py]: the number of bins should be added as a parameter to `data_loss_histogram`.

A user might further want to customize the appearance of the bars, so it also might be worth adding `color`, `edgecolor` and `alpha`.

File: src/pymovements/plotting/data_loss_histogram.py (hunk 9 above)
Comment (dkrako) [inline src/pymovements/plotting/data_loss_histogram.py]: the ticks on the x axis are misaligned which looks a bit strange. It's not really clear what duration they represent exactly. placing the ticks at exactly the center of the bars might solve that issue (in case of single-element bins this would be the exact value, in case of wider bins this would represent the mean).

File: src/pymovements/plotting/data_loss_histogram.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+        ax.set_xlabel('Chunk Length' + (' (ms)' if unit == 'time' else ' (samples)'))
+        ax.set_ylabel('Frequency')
+
```
Comment (dkrako) [inline src/pymovements/plotting/data_loss_histogram.py:197]: As the count will always be a non-negative integer the ticks on the y axis should not include fractions as this might be confusing.

**Review by dkrako — CHANGES_REQUESTED** [review]
Thank you so much for this new plotting function! Sorry for the delay, I was on vacation and it took me some time until I got to review this.

I have some minor change requests which should mostly be quick and easy to apply. The issue with the x-axis ticks might be a bit tricky to get right, but I think it's important for being able to correctly interpret the plot.

Also please make sure to resolve all the errors that are raised by pre-commit and pylint.

EDIT: The weird failing test in `tests/unit/gaze/transforms/resample_test.py` is unrelated to your PR and is already resolved on main, so updating your branch should get rid of that.

**Comment by dkrako** [comment]
@Helly121 Also: feel encouraged to add yourself to the list of authors in our [CITATION.cff](https://github.com/pymovements/pymovements/blob/main/CITATION.cff).

File: docs/source/user-guide/inspect-raw-samples.ipynb (hunk 10 above)
Comment (Helly121) [reply docs/source/user-guide/inspect-raw-samples.ipynb]: You are right, the histogram bars were all similar heights (1-2) because each gap length appeared the same number of times. I've updated the CSV so short gaps appear more frequently.

**Comment by Helly121** [comment]
Sorry for such long delay, I had my college end semester exams and wasn't able to contribute here, I have addressed all the changes mentioned by @dkrako. I noticed that after i updated my branch there was a huge change in the main branch, I assume there was a lot of re-structuring while I was not around? 
Either way please do let me know if I missed anything, Thanks.

**Comment by Helly121** [comment]
Hello, what needs to be done more for the PR to be merged?

**Review by dkrako — APPROVED** [review]
Hi @Helly121, sorry for keeping you waiting so long. I was on and off vacations and a conference in the last months.

I updated your PR but I was unsure if some of the logic couldn't be refactored to use pymovements segmentation transforms (`segmentation2events()`). However I think that the implementation is fine so I'll merge it as is as there are currently many PRs open. If you'd like the challenge, feel free to checkout how this could be implemented with pymovements functions.

Thanks again for your excellent work! Feel encouraged to create a follow-up PR for your other plotting ideas!

**Comment by Helly121** [comment]
Thanks a lot for merging, I will certainly look into the other plotting functitons
